File: frontend/src/game/ai.ts
import { BOARD_SIZE } from './constants';
import { coordKey, inBounds } from './board';
import type { Coord, ShotResult } from './types';
import type { Rng } from './board';

/**
 * Difficulty levels for the AI opponent.
 * - `easy`   : fires completely at random.
 * - `medium` : hunt/target — random search, then focuses around hits.
 * - `hard`   : medium plus a checkerboard "parity" search, which is the
 *              optimal way to find ships of length >= 2 with fewer shots.
 */
export type Difficulty = 'easy' | 'medium' | 'hard';

/**
 * Mutable memory the AI carries between shots. It only ever records the
 * results of its *own* shots — it never reads the opponent's ship positions,
 * so it plays fairly.
 */
export interface AIState {
  difficulty: Difficulty;
  /** Coordinate keys the AI has already fired at. */
  tried: Set<string>;
  /** Promising cells to try next (neighbours of unresolved hits). */
  targetQueue: Coord[];
  /** Hits belonging to the ship currently being hunted, in fire order. */
  activeHits: Coord[];
}

export function createAI(difficulty: Difficulty = 'hard'): AIState {
  return { difficulty, tried: new Set(), targetQueue: [], activeHits: [] };
}

const NEIGHBOURS: Coord[] = [
  { row: -1, col: 0 },
  { row: 1, col: 0 },
  { row: 0, col: -1 },
  { row: 0, col: 1 },
];

function untriedNeighbours(state: AIState, c: Coord): Coord[] {
  return NEIGHBOURS.map((d) => ({ row: c.row + d.row, col: c.col + d.col }))
    .filter(inBounds)
    .filter((n) => !state.tried.has(coordKey(n)));
}

/**
 * If the AI has two or more in-line hits on the active ship, return the two
 * cells that extend that line (so it finishes ships efficiently instead of
 * poking randomly around a hit).
 */
function lineExtensions(state: AIState): Coord[] {
  const hits = state.activeHits;
  if (hits.length < 2) return [];

  const sameRow = hits.every((h) => h.row === hits[0].row);
  const sameCol = hits.every((h) => h.col === hits[0].col);
  if (!sameRow && !sameCol) return [];

  const extensions: Coord[] = [];
  if (sameRow) {
    const cols = hits.map((h) => h.col);
    const row = hits[0].row;
    extensions.push({ row, col: Math.min(...cols) - 1 });
    extensions.push({ row, col: Math.max(...cols) + 1 });
  } else {
    const rows = hits.map((h) => h.row);
    const col = hits[0].col;
    extensions.push({ row: Math.min(...rows) - 1, col });
    extensions.push({ row: Math.max(...rows) + 1, col });
  }
  return extensions.filter(inBounds).filter((c) => !state.tried.has(coordKey(c)));
}

/** All cells that have never been fired at. */
function availableCells(state: AIState): Coord[] {
  const cells: Coord[] = [];
  for (let row = 0; row < BOARD_SIZE; row++) {
    for (let col = 0; col < BOARD_SIZE; col++) {
      const c = { row, col };
      if (!state.tried.has(coordKey(c))) cells.push(c);
    }
  }
  return cells;
}

function pick<T>(arr: T[], rng: Rng): T {
  return arr[Math.floor(rng() * arr.length)];
}

/**
 * Choose the AI's next target. Pure with respect to game state — the only
 * external input is the RNG, which makes behaviour deterministic in tests.
 */
export function chooseMove(state: AIState, rng: Rng = Math.random): Coord {
  // 1. Finish a wounded ship by extending the line of hits.
  if (state.difficulty !== 'easy') {
    const extensions = lineExtensions(state);
    if (extensions.length > 0) return pick(extensions, rng);

    // 2. Otherwise probe queued neighbours of an isolated hit. We *find* rather
    //    than mutate the queue so the function stays pure (no side effects on
    //    the caller's state); stale entries are skipped via the `tried` guard.
    const queued = state.targetQueue.find(
      (c) => inBounds(c) && !state.tried.has(coordKey(c)),
    );
    if (queued) return queued;
  }

  const available = availableCells(state);

  // 3. Hard mode: search on a checkerboard so we never waste shots on cells
  //    that cannot complete a 2-cell ship until the parity set is exhausted.
  if (state.difficulty === 'hard') {
    const parityCells = available.filter((c) => (c.row + c.col) % 2 === 0);
    if (parityCells.length > 0) return pick(parityCells, rng);
  }

  return pick(available, rng);
}

/**
 * Record the result of a shot and return updated AI memory. Treated immutably:
 * the returned state is a fresh object.
 */
export function registerResult(
  state: AIState,
  target: Coord,
  result: ShotResult,
): AIState {
  if (result === 'already') return state;

  const tried = new Set(state.tried);
  tried.add(coordKey(target));

  let activeHits = state.activeHits;
  let targetQueue = state.targetQueue;

  if (result === 'hit') {
    activeHits = [...state.activeHits, target];
    targetQueue = [...state.targetQueue, ...untriedNeighbours({ ...state, tried }, target)];
  } else if (result === 'sunk') {
    // The wounded ship is finished — forget it and resume hunting from scratch.
    activeHits = [];
    targetQueue = [];
  }

  return { ...state, tried, activeHits, targetQueue };
}

File: backend/app/game/ai.py
"""Server-side Battleship AI.

This is a faithful port of the client's ``ai.ts`` so the backend can serve the
opponent's moves. The AI only ever reasons about the results of its *own*
shots (which the client passes in), so it never cheats by reading ship
positions.

The board is 10x10. Coordinates are ``(row, col)`` tuples, zero-indexed.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field

BOARD_SIZE = 10

Coord = tuple[int, int]

Difficulty = str  # one of: "easy", "medium", "hard"

_NEIGHBOURS: tuple[Coord, ...] = ((-1, 0), (1, 0), (0, -1), (0, 1))


def in_bounds(c: Coord) -> bool:
    """Whether a coordinate lies on the board."""
    row, col = c
    return 0 <= row < BOARD_SIZE and 0 <= col < BOARD_SIZE


@dataclass
class AIMemory:
    """Everything the AI remembers between shots.

    Mirrors the client's ``AIState``. ``tried`` holds every cell already fired
    at, ``active_hits`` are the unresolved hits on the ship currently being
    hunted, and ``target_queue`` are promising neighbours to try next.
    """

    difficulty: Difficulty = "hard"
    tried: set[Coord] = field(default_factory=set)
    active_hits: list[Coord] = field(default_factory=list)
    target_queue: list[Coord] = field(default_factory=list)


def _line_extensions(mem: AIMemory) -> list[Coord]:
    """Cells that extend a straight line of two-or-more in-line hits."""
    hits = mem.active_hits
    if len(hits) < 2:
        return []

    same_row = all(h[0] == hits[0][0] for h in hits)
    same_col = all(h[1] == hits[0][1] for h in hits)
    if not same_row and not same_col:
        return []

    extensions: list[Coord] = []
    if same_row:
        row = hits[0][0]
        cols = [h[1] for h in hits]
        extensions.append((row, min(cols) - 1))
        extensions.append((row, max(cols) + 1))
    else:
        col = hits[0][1]
        rows = [h[0] for h in hits]
        extensions.append((min(rows) - 1, col))
        extensions.append((max(rows) + 1, col))

    return [c for c in extensions if in_bounds(c) and c not in mem.tried]


def _available_cells(mem: AIMemory) -> list[Coord]:
    return [
        (r, c)
        for r in range(BOARD_SIZE)
        for c in range(BOARD_SIZE)
        if (r, c) not in mem.tried
    ]


def choose_move(mem: AIMemory, rng: random.Random | None = None) -> Coord:
    """Choose the AI's next target. Pure aside from the supplied RNG."""
    rng = rng or random

    if mem.difficulty != "easy":
        # 1. Finish a wounded ship by extending the hit line.
        extensions = _line_extensions(mem)
        if extensions:
            return rng.choice(extensions)

        # 2. Probe queued neighbours of an isolated hit (skip stale entries).
        queued = [c for c in mem.target_queue if in_bounds(c) and c not in mem.tried]
        if queued:
            return queued[0]

    available = _available_cells(mem)
    if not available:
        raise ValueError("no available cells remain")

    # 3. Hard mode: search a checkerboard to find ships in the fewest shots.
    if mem.difficulty == "hard":
        parity = [c for c in available if (c[0] + c[1]) % 2 == 0]
        if parity:
            return rng.choice(parity)

    return rng.choice(available)


def register_result(mem: AIMemory, target: Coord, result: str) -> AIMemory:
    """Return updated memory after observing the result of ``target``.

    ``result`` is one of ``"miss"``, ``"hit"``, ``"sunk"`` or ``"already"``.
    """
    if result == "already":
        return mem

    mem.tried.add(target)

    if result == "hit":
        mem.active_hits.append(target)
        for dr, dc in _NEIGHBOURS:
            n = (target[0] + dr, target[1] + dc)
            if in_bounds(n) and n not in mem.tried:
                mem.target_queue.append(n)
    elif result == "sunk":
        mem.active_hits = []
        mem.target_queue = []

    return mem
